### PR TITLE
Prepare split of output walkers and tree walkers

### PR DESCRIPTION
**Pull request by derrabus**
Prepares #9551. Please see that PR for details & discussion.

### CODE REVIEW

File: UPGRADE.md [hunk 1]
```diff
@@ -1,5 +1,72 @@
...
+* `walkSimpleArithmeticExpression()`
+* `walkPathExpression()`
+* `walkResultVariable()`
+* `getExecutor()`
+
+The following changes have been made to the abstract `TreeWalkerAdapter` class:
+
+* All implementations of now-deprecated `TreeWalker` methods have been
+  deprecated as well.
+* The method `setQueryComponent()` will become protected in 3.0. Calling is
```
Comment (greg0ire) [inline UPGRADE.md]:
```suggestion
* The method `setQueryComponent()` will become protected in 3.0. Calling it
```

File: tests/Doctrine/Tests/ORM/Query/TreeWalkerAdapterTest.php [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+class TreeWalkerAdapterTest extends TestCase
+{
+    use VerifyDeprecations;
+
+    public function testDeprecatedSetQueryComponent(): void
+    {
+        $walker = new class (
+            $this->createMock(AbstractQuery::class),
+            $this->createMock(ParserResult::class),
```
Comment (greg0ire) [inline tests/Doctrine/Tests/ORM/Query/TreeWalkerAdapterTest.php:24]:
```suggestion
            $this->createStub(AbstractQuery::class),
            $this->createStub(ParserResult::class),
```

File: tests/Doctrine/Tests/ORM/Query/TreeWalkerAdapterTest.php [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+            'nestingLevel' => 0,
+            'token' => ['value' => '', 'type' => Lexer::T_NONE, 'position' => 0],
+        ]);
+    }
+
+    public function testSetQueryComponent(): void
+    {
+        $walker = new class (
+            $this->createMock(AbstractQuery::class),
+            $this->createMock(ParserResult::class),
```
Comment (greg0ire) [inline tests/Doctrine/Tests/ORM/Query/TreeWalkerAdapterTest.php:44]:
```suggestion
            $this->createStub(AbstractQuery::class),
            $this->createStub(ParserResult::class),
```

**Comment by greg0ire** [comment]
Sorry if I reviewed this a bit too early, I'll wait for you to request my review next time :P

**Comment by derrabus** [comment]
No, please go ahead.

File: lib/Doctrine/ORM/Query/TreeWalkerAdapter.php [hunk 4]
```diff
@@ -22,60 +27,90 @@ abstract class TreeWalkerAdapter implements TreeWalker
...
+     *
+     * @return void
      */
     public function setQueryComponent($dqlAlias, array $queryComponent)
     {
+        $trace = debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS, 2);
+        if (! isset($trace[1]['class']) || ! is_a($trace[1]['class'], self::class, true)) {
+            Deprecation::trigger(
+                'doctrine/orm',
+                'TODO',
```
Comment (greg0ire) [inline lib/Doctrine/ORM/Query/TreeWalkerAdapter.php]: Sure! :point_up: 

:trollface:

Comment (derrabus) [reply lib/Doctrine/ORM/Query/TreeWalkerAdapter.php]: 😖

Comment (greg0ire) [reply lib/Doctrine/ORM/Query/TreeWalkerAdapter.php]: Haha we need to setup a pre-commit hook for this :smile:

**Comment by greg0ire** [comment]
Looks like `createStub` was a bad idea… that's weird to me, cause we already have some in the codebase :thinking: 

https://github.com/doctrine/orm/blob/48e4e333c7bf4757e9a4e19416d1a3e833ea28b0/tests/Doctrine/Tests/ORM/ConfigurationTest.php#L160

I always forget if we can use it or not, so I checked the codebase beforehand. Maybe that test is skipped somehow.

**Review by greg0ire — APPROVED** [review]

**Pull request merged**